### DISCUSSION TITLE
Nominal and other clock specs on a per core basis?

### DISCUSSION THREAD

**Discussion by GerhardPehland** [opcm/pcm#609] [Q&A]
With the arrival of performance and effiency cores the various clock specs are different for different cores, Is there a way to determine these values on a per core basis with the Intel PCM library?

**Comment by opcm** [comment]
AFAIK the nominal frequency is the same across P and E cores on hybrid processors. It can be read from the 0xCE MSR: 

```
pcm-msr.exe -a 0xCE -b 8:15 -d

 Intel(r) Performance Counter Monitor ($Format:%ci ID=%h$)

 MSR read/write utility

=====  Processor information  =====
Hybrid processor         : yes
...
Nominal core frequency: 2500000000 Hz

Socket 0: 0 PCU units detected. 0 IIO units detected. 0 IRP units detected. 0 CHA/CBO units detected. 0 MDF units detected. 0 CXL units detected.

 Read bits 8:15 value 25 from MSR 206 on core 0

 Read bits 8:15 value 25 from MSR 206 on core 1

 Read bits 8:15 value 25 from MSR 206 on core 2

 Read bits 8:15 value 25 from MSR 206 on core 3

 Read bits 8:15 value 25 from MSR 206 on core 4

 Read bits 8:15 value 25 from MSR 206 on core 5

 Read bits 8:15 value 25 from MSR 206 on core 6

 Read bits 8:15 value 25 from MSR 206 on core 7

 Read bits 8:15 value 25 from MSR 206 on core 8

 Read bits 8:15 value 25 from MSR 206 on core 9

 Read bits 8:15 value 25 from MSR 206 on core 10

 Read bits 8:15 value 25 from MSR 206 on core 11

 Read bits 8:15 value 25 from MSR 206 on core 12

 Read bits 8:15 value 25 from MSR 206 on core 13

 Read bits 8:15 value 25 from MSR 206 on core 14

 Read bits 8:15 value 25 from MSR 206 on core 15

```

25 * 100 000 000 nominal frequency is the same on all cores on my hybrid Alderlake system

**Reply by GerhardPehland** [reply]
Hmm. I am also reading MSR 1AD hex, which should give me the max turbo multipliuers. But I get the same value for all cores, E and P. That can't be right? Or am I going about it in the wrong way?

What I am doing is:

SafeMsrHandle msr(c); // c from 0 to the number of cores
uint64 nReg1AD;
msr.read(0x1ad, &nReg1AD);

The value I read is the same for all cores. I would have expected different values for P an E cores. The value I get translates to a max boost frequency of 6 GHz on an I9-14900K with one core active.

**Reply by GerhardPehland** [reply]
Found out that reading register 1AD from different cores is not the correct way.. The value is for the package, not a specific core.

I did find MSR 650H, which supplies a "secondary" table of turbo multipliers for the efficiency cores. It supplies a max frequency of 4.4 GHx (multiplier 44) for the I9-14900K, which is correct.

I have yet to find an MSR that supplies the  "Maximum non-turbo ratio" for the E cores, I would expect a value of 24 (2.4 Ghz) for the 14900K. Also the  "Maximum efficiency ratio" clock speed on E-cores. For the P cores these two values are found in MSR CEH, but I cannot find an equivalent for E cores.

Any ideas anyone?
